homeassistant/components/geofency/__init__.py

```diff
@@ -50,97 +48,107 @@
-
-        for attribute in gps_attributes:
-            if attribute in data:
-                data[attribute] = float(data[attribute])
-
-        return data
-
-    def _is_mobile_beacon(self, data):
-        """Check if we have a mobile beacon."""
-        return 'beaconUUID' in data and data['name'] in self.mobile_beacons
-
-    @staticmethod
-    def _device_name(data):
-        """Return name of device tracker."""
-        if 'beaconUUID' in data:
-            return "{}_{}".format(BEACON_DEV_PREFIX, data['name'])
-        return data['device']
-
-    async def _set_location(self, hass, data, location_name):
-        """Fire HA event to set location."""
-        device = self._device_name(data)
-
-        async_dispatcher_send(
-            hass,
-            TRACKER_UPDATE,
-            device,
-            (data[ATTR_LATITUDE], data[ATTR_LONGITUDE]),
-            location_name,
-            data
-        )
-
-        return "Setting location for {}".format(device)
+async def handle_webhook(hass, webhook_id, request):
+    """Handle incoming webhook with Mailgun inbound messages."""
+    data = _validate_data(await request.post())
+
+    if not data:
+        return "Invalid data", HTTP_UNPROCESSABLE_ENTITY
+
+    if _is_mobile_beacon(data, hass.data[DOMAIN]):
+        return await _set_location(hass, data, None)
+    if data['entry'] == LOCATION_ENTRY:
+        location_name = data['name']
+    else:
+        location_name = STATE_NOT_HOME
+        if ATTR_CURRENT_LATITUDE in data:
+            data[ATTR_LATITUDE] = data[ATTR_CURRENT_LATITUDE]
+            data[ATTR_LONGITUDE] = data[ATTR_CURRENT_LONGITUDE]
+
+    return await _set_location(hass, data, location_name)
+
+
+def _validate_data(data):
+    """Validate POST payload."""
+    data = data.copy()
+
+    required_attributes = ['address', 'device', 'entry',
+                           'latitude', 'longitude', 'name']
+
+    valid = True
+    for attribute in required_attributes:
+        if attribute not in data:
+            valid = False
+            _LOGGER.error("'%s' not specified in message", attribute)
+
+    if not valid:
+        return {}
+
+    data['address'] = data['address'].replace('\n', ' ')
+    data['device'] = slugify(data['device'])
+    data['name'] = slugify(data['name'])
+
+    gps_attributes = [ATTR_LATITUDE, ATTR_LONGITUDE,
+                      ATTR_CURRENT_LATITUDE, ATTR_CURRENT_LONGITUDE]
+
+    for attribute in gps_attributes:
+        if attribute in data:
+            data[attribute] = float(data[attribute])
+
+    return data
+
+
+def _is_mobile_beacon(self, data):
+    """Check if we have a mobile beacon."""
+    return 'beaconUUID' in data and data['name'] in self.mobile_beacons
+
+
+def _device_name(data):
+    """Return name of device tracker."""
+    if 'beaconUUID' in data:
+        return "{}_{}".format(BEACON_DEV_PREFIX, data['name'])
+    return data['device']
+
+
+async def _set_location(hass, data, location_name):
```

Comment:
no need to be async? 

Also maybe no need to be it's own function ?

Reply:
It's called twice though, seems worth keeping it as a separate function.